### ISSUE TITLE
Free SSL with CloudFlare and Heroku

### ISSUE THREAD

**Issue by winston** [jollygoodcode/jollygoodcode.github.io#9]
Recently, I set up CloudFlare with Heroku to make good use of its [Universal SSL](https://www.cloudflare.com/ssl) and essentially made [Dasherize](https://www.dasherize.com) `https` the poor man's way.

My aim was to get the following working:
- `http://dasherize.com` redirects to `https://www.dasherize.com`
- `http://www.dasherize.com` redirects to `https://www.dasherize.com`
- `https://dasherize.com` redirects to `https://www.dasherize.com`
- `https://www.dasherize.com` works!

Here are the steps to get that working:
### 1) Sign up for a CloudFlare account

Go to https://www.cloudflare.com/.
### 2) Add a website to CloudFlare

![1](https://cloud.githubusercontent.com/assets/2112/11184357/fd84a10c-8cb1-11e5-859f-6c558eb8ddfa.png)
### 3) Configure CNAME

After you have scanned your website, you will probably see an `A` entry and a `CNAME` entry.

Modify (and/or delete) the `A` and `CNAME` entries so that they become:
- `CNAME`, `dasherize.com` to Heroku domain name
- `CNAME`, `www` to Heroku domain name

It might look strange to have two `CNAME` going to the same Heroku domain, but CloudFlare supports [CNAME Flattening](https://support.cloudflare.com/hc/en-us/articles/200169056-CNAME-Flattening-RFC-compliant-support-for-CNAME-at-the-root) so we are good.

![2](https://cloud.githubusercontent.com/assets/2112/11184367/0c87e380-8cb2-11e5-9160-33124a6b745f.png)
### 4) Wait for DNS to propagate

At this point, we can wait for DNS to propagate and when it's done:
- `http://dasherize.com` redirects to `https://dasherize.com`
- `http://www.dasherize.com` redirects to `https://www.dasherize.com`

And the DNS entries should look like so (with some information ommitted):

```
$ curl -I http://dasherize.com
HTTP/1.1 301 Moved Permanently
...
Location: https://dasherize.com/
Via: 1.1 vegur
Server: cloudflare-nginx

$ curl -I http://www.dasherize.com
HTTP/1.1 301 Moved Permanently
...
Location: https://www.dasherize.com/
Via: 1.1 vegur
Server: cloudflare-nginx

$ curl -I https://dasherize.com
HTTP/1.1 200 OK
Server: cloudflare-nginx
...
Via: 1.1 vegur

$ curl -I https://www.dasherize.com
HTTP/1.1 200 OK
Server: cloudflare-nginx
..
Via: 1.1 vegur
```

We are almost there, we are just left with redirecting `http://dasherize.com` to `https://www.dasherize.com`.

You might be thinking.. But why `www`? [Everyone has different opinions](http://www.sitepoint.com/domain-www-or-no-www/).
### 5) Final Redirection

To redirect `http://dasherize.com` to `https://www.dasherize.com`, we need to include add a `Page Rule` that:

![3](https://cloud.githubusercontent.com/assets/2112/11184499/031025aa-8cb3-11e5-8f70-58f2dfe7cd65.png)

> Forwards (301) `https://dasherize.com/*` to `https://www.dasherize.com`

And with that, the DNS entries will look like:

```
$ curl -I http://dasherize.com
HTTP/1.1 301 Moved Permanently
...
Location: https://dasherize.com/
Via: 1.1 vegur
Server: cloudflare-nginx

$ curl -I http://www.dasherize.com
HTTP/1.1 301 Moved Permanently
...
Location: https://www.dasherize.com/
Via: 1.1 vegur
Server: cloudflare-nginx

$ curl -I https://dasherize.com
HTTP/1.1 301 Moved Permanently
...
Server: cloudflare-nginx
Location: https://www.dasherize.com/

$ curl -I https://www.dasherize.com
HTTP/1.1 200 OK
...
Server: cloudflare-nginx
Via: 1.1 vegur
```
### 6) Full SSL

Finally, go to the `Crypto` page, and make sure that you have selected the `Full` option for your SSL. You can read more about the differences by clicking on `Help` below the select options.

![screen shot 2015-11-16 at 10 35 20 pm](https://cloud.githubusercontent.com/assets/2112/11184505/0c9deb8e-8cb3-11e5-8523-68e3ee0ac516.png)

---

With these 6 steps, you now have a SSL enabled site for $0, all thanks to CloudFlare's Full SSL option:

![4](https://cloud.githubusercontent.com/assets/2112/11184396/357111ea-8cb2-11e5-912e-323029ab59d1.png)

Since all Heroku apps comes free with `https` and that, quoting CloudFlare, "CloudFlare will not attempt to validate the certificate", hence it makes it easy for us to have the Dasherize site SSL-enabled.

Also, don't forget to set `config.force_ssl = true` in your Rails `production.rb`.

---

Thank you for reading.

@winston :pencil2: [Jolly Good Code](http://www.jollygoodcode.com)
### About Jolly Good Code

[![Jolly Good Code](https://cloud.githubusercontent.com/assets/1000669/9362336/72f9c406-46d2-11e5-94de-5060e83fcf83.jpg)](http://www.jollygoodcode.com)

We specialise in Agile practices and Ruby, and we love contributing to open source. 
[Speak to us](http://www.jollygoodcode.com/#get-in-touch) about your next big idea, or [check out our projects](http://www.jollygoodcode.com/open-source).

**Comment by joshteng** [comment]
@winston tremendously helpful and cost saving tip!

**Comment by winston** [comment]
@joshteng ka-ching!

**Comment by fadhlirahim** [comment]
:+1:

**Comment by tmlee** [comment]
@winston Nice that does the job in a few steps! 
I supposed for something more granular, https://github.com/tobmatth/rack-ssl-enforcer can be considered?

**Comment by winston** [comment]
@tmlee Yup. That gem seems to do the job.

**Comment by faizalzakaria** [comment]
interesting.

**Comment by lordhumunguz** [comment]
Thanks Winston!

the redirecting from http to https doesn't seem to be automatic. Did you have to turn it on somewhere? I get 200 OK's for both http and https.

**Comment by joshteng** [comment]
@4thethrillofit I do a 301 http -> https redirection at my DNS. fyi, using cloudflare. But I think if ure still using Rails, there's a config to force ssl I think.

**Comment by winston** [comment]
@4thethrillofit Is it a Rails app? As what @joshteng said:

> Also, don't forget to set config.force_ssl = true in your Rails production.rb.

**Comment by lordhumunguz** [comment]
Yep! @joshteng apparently they want you to use [Page Rules](https://support.cloudflare.com/hc/en-us/articles/200170536-How-do-I-redirect-all-visitors-to-HTTPS-SSL-)

**Comment by joshteng** [comment]
@4thethrillofit it was weird at first but i got to love it fast enough. haha

**Comment by choonkeat** [comment]
another day, another free SSL https://aws.amazon.com/blogs/aws/new-aws-certificate-manager-deploy-ssltls-based-apps-on-aws/

> ... at no extra cost. SSL/TLS certificates provisioned through AWS Certificate Manager are free!
> 
> ACM will allow you to start using SSL in a matter of minutes. After your request a certificate, you can deploy it to your Elastic Load Balancers and your Amazon CloudFront distributions with a couple of clicks. After that, ACM can take care of the periodic renewals without any action on your part.

**Comment by guptarohit** [comment]
Is dasherize.com hosting on heroku?
I am using shared hosting, how can I make it work.
In my case records looks like(I have removed 'A' records):

| Type | Name | Value | TTL |
| --- | --- | --- | --- |
| CNAME | mydomain.xyz | is an alias of clcp.hostingprovider.com | Automatic |
| CNAME | www | is an alias of clcp.hostingprovider.com | Automatic |

still not working, I also tried the value of A record(which I've removed now) in place of CNAME's value. Can you suggest possible solution?

**Comment by winston** [comment]
@guptarohit Yes dasherize is hosted on heroku.

Is clcp.hostingprovider.com your actual application?

Is it because of a DNS propagation issue? Does it work already?

**Comment by guptarohit** [comment]
@winston no clcp.hostingprovider.com is not my application.
My application is hosted by a hosting provider.
"clcp.hostingprovider.com" got inserted in CNAME while I setting up cloudflare. I'd just pointed my DNS to cloudflare's nameserver, that's all I did.
There was some IP address in A record(which I have removed).

**Comment by winston** [comment]
Hmm.. First question.. does your hosting provider provide you with a https end point? If not, it's pretty futile to follow this guide. i.e. it only works with Heroku.

Next, is your application accessible via an IP address or URL? Try replacing "clcp.hostingprovider.com" with that. But again, if your app is not accessible at `https://<ip or url`> on your hosting provider, like I said, this won't work. Hope this clarifies.

**Comment by guptarohit** [comment]
Exactly this is the problem, <code>https://&lt;ip or url</code> does not resolve to my application.
:+1: Thanks though for clearing my query. :smiley:
